examples/Locomotion/Basic_Thumbsticks/main.lua

```diff
@@ -0,0 +1,119 @@
+-- Left thumbstick: move and strafe
+-- Right thumbstick: rotate the view horizontally
+
+-- Right grip: switch to snap locomotion mode
+-- Left grip: enable flying mode
+-- Left trigger: switch orientation to left hand controller (can look around while moving)
+
+local motion = {
+  pose = lovr.math.newMat4(), -- Transformation in VR initialized to origin (0,0,0) looking down -Z
+  thumbstickDeadzone = 0.4,   -- Smaller thumbstick displacements are ignored (too much noise)
+  flying = false,
+  directionFrom = 'head',     -- Movement can be relative to orientation of head or left controller
+  -- Snap motion parameters
+  snapTurnAngle = 2 * math.pi / 12,
+  dashDistance = 1.5,
+  thumbstickCooldownTime = 0.3,
+  thumbstickCooldown = 0,
+  -- Smooth motion parameters
+  turningSpeed = 2 * math.pi * 1 / 6,
+  walkingSpeed = 4,
+}
+
+function motion.smooth(dt)
+  if lovr.headset.isTracked('right') then
+    local x, y = lovr.headset.getAxis('right', 'thumbstick')
+    if math.abs(x) > motion.thumbstickDeadzone then
+      motion.pose:rotate(-x * motion.turningSpeed * dt, 0, 1, 0)
+    end
+  end
+  if lovr.headset.isTracked('left') then
+    local x, y = lovr.headset.getAxis('left', 'thumbstick')
+    local direction = quat(lovr.headset.getOrientation(motion.directionFrom)):direction()
+    if not motion.flying then
+      direction.y = 0
+    end
+    if math.abs(x) > motion.thumbstickDeadzone then
+      local strafeVector = quat(-math.pi / 2, 0,1,0):mul(vec3(direction))
+      motion.pose:translate(strafeVector * x * motion.walkingSpeed * dt)
+    end
+    if math.abs(y) > motion.thumbstickDeadzone then
+      motion.pose:translate(direction * y * motion.walkingSpeed * dt)
+    end
+  end
+end
+
+function motion.snap(dt)
+  if lovr.headset.isTracked('right') then
+    local x, y = lovr.headset.getAxis('right', 'thumbstick')
+    if math.abs(x) > motion.thumbstickDeadzone and motion.thumbstickCooldown < 0 then
+      local direction = -x / math.abs(x)
+      motion.pose:rotate(motion.snapTurnAngle * direction, 0, 1, 0)
+      motion.thumbstickCooldown = motion.thumbstickCooldownTime
+    end
+  end
+  if lovr.headset.isTracked('left') then
+    local x, y = lovr.headset.getAxis('left', 'thumbstick')
+    if math.abs(y) > motion.thumbstickDeadzone and motion.thumbstickCooldown < 0 then
+      local direction = quat(lovr.headset.getOrientation('head')):direction()
+      if not motion.flying then
+        direction.y = 0
+      end
+      local moveVector = direction * motion.dashDistance * y / math.abs(y)
+      motion.pose:translate(moveVector)
+      motion.thumbstickCooldown = motion.thumbstickCooldownTime
+    end
+  end
+  motion.thumbstickCooldown = motion.thumbstickCooldown - dt
+end
+
+function lovr.update(dt)
```

Comment:
If this was `motion.update`, then the `motion` table would be a small self-contained copy-pastable library people could use.

Reply:
The `motion` table was intentionally isolated for this purpose and I think I drew the line correctly. I would prefer the user to control the update logic. The controller mappings make sense only to expose all options and they use up most useful buttons on controllers. The flying / grounded mechanics is fairly specific to plane-restricted movement and should be part of game logic.

When other locomotion examples are done (teleporting, jogging, swimming), it will become more obvious how the `motion` library should expose different options. Users should be able to compose update and rendering functions to mix and match between modes of locomotion.